### PR TITLE
netutils/dhcpc: add non-blocking interface

### PR DESCRIPTION
**Pull request by coffee809721232**
## Summary
netutils/dhcpc: add non-blocking interface

Signed-off-by: songlinzhang <songlinzhang@xiaomi.com>
## Impact
Add dhcpc_request_async function
## Testing
Replace dhcpc_request

### CODE REVIEW

File: netutils/dhcpc/dhcpc.c [hunk 1]
```diff
@@ -475,8 +520,8 @@ FAR void *dhcpc_open(FAR const char *interface, FAR const void *macaddr,
 
       memset(pdhcpc, 0, sizeof(struct dhcpc_state_s));
       pdhcpc->interface  = interface;
-      pdhcpc->ds_macaddr = macaddr;
-      pdhcpc->ds_maclen  = maclen;
+      pdhcpc->ds_maclen  = maclen > IFHWADDRLEN ? IFHWADDRLEN : maclen;
```
Comment (hartmannathan) [inline netutils/dhcpc/dhcpc.c:523]: If `maclen` too large, should we return error status (NULL) instead of proceeding?

**Review by hartmannathan — COMMENTED** [review]
LGTM (with just the one question about `maclen` in dhcpc_open())

File: netutils/dhcpc/dhcpc.c [hunk 2]
```diff
@@ -468,15 +512,15 @@ FAR void *dhcpc_open(FAR const char *interface, FAR const void *macaddr,
 
   /* Allocate an internal DHCP structure */
 
-  pdhcpc = (FAR struct dhcpc_state_s *)malloc(sizeof(struct dhcpc_state_s));
+  pdhcpc = malloc(sizeof(struct dhcpc_state_s) + maclen + 1);
```
Comment (xiaoxiang781216) [inline netutils/dhcpc/dhcpc.c:515]: change + 1 to - 1

File: netutils/dhcpc/dhcpc.c (hunk 1 above)
Comment (coffee809721232) [reply netutils/dhcpc/dhcpc.c]: I adjusted the storage method of mac address, please review.

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

**Comment by hartmannathan** [comment]
Looks like I'm late to the party but LGTM.

Just one question:

Are we sure we want to remove the typecast from the (void *) returned by malloc() to (FAR struct dhcpc_state_s *) in this call:

```
-pdhcpc = (FAR struct dhcpc_state_s *)malloc(sizeof(struct dhcpc_state_s));
+pdhcpc = malloc(sizeof(struct dhcpc_state_s) + maclen - 1);
```

Some compilers may warn about that.

Might also benefit from a comment explaining the `maclen - 1` such as:

```
/* maclen - 1: First byte of maclen is included in the structure */
```

So that someone in the future won't think it's a mistake.

**Comment by xiaoxiang781216** [comment]
> Looks like I'm late to the party but LGTM.
> 
> Just one question:
> 
> Are we sure we want to remove the typecast from the (void *) returned by malloc() to (FAR struct dhcpc_state_s *) in this call:
> 
> ```
> -pdhcpc = (FAR struct dhcpc_state_s *)malloc(sizeof(struct dhcpc_state_s));
> +pdhcpc = malloc(sizeof(struct dhcpc_state_s) + maclen - 1);
> ```
> 

It's a normal practice to let compiler auto convert void * to type * in C:
https://stackoverflow.com/questions/605845/do-i-cast-the-result-of-malloc

> Some compilers may warn about that.
> 

This implicit conversion is allowed by C standard. But, it forbid by C++.


> Might also benefit from a comment explaining the `maclen - 1` such as:
> 
> ```
> /* maclen - 1: First byte of maclen is included in the structure */
> ```
> 
> So that someone in the future won't think it's a mistake.

Yes, good suggestion.

**Comment by hartmannathan** [comment]
> > Are we sure we want to remove the typecast from the (void *) returned by malloc() to (FAR struct dhcpc_state_s *) in this call:
> > ```
> > -pdhcpc = (FAR struct dhcpc_state_s *)malloc(sizeof(struct dhcpc_state_s));
> > +pdhcpc = malloc(sizeof(struct dhcpc_state_s) + maclen - 1);
> > ```
> 
> It's a normal practice to let compiler auto convert void * to type * in C: https://stackoverflow.com/questions/605845/do-i-cast-the-result-of-malloc
> 
> > Some compilers may warn about that.
> 
> This implicit conversion is allowed by C standard. But, it forbid by C++.

Ah yes, that's what it is. In my own work I code C so it builds as both C and C++, so I forgot that's a C++ warning. Here I guess it's okay since we are coding for C89.

Thanks!

File: netutils/dhcpc/dhcpc.c [hunk 3]
```diff
@@ -445,6 +448,47 @@ static uint8_t dhcpc_parsemsg(FAR struct dhcpc_state_s *pdhcpc, int buflen,
...
+        }
+
+      if (pdhcpc->cancel)
+        {
+          return NULL;
+        }
+
+      while (result.lease_time)
+        {
+          result.lease_time = sleep(result.lease_time);
```
Comment (normanr) [inline netutils/dhcpc/dhcpc.c:481]: Note: This should only sleep half of the lease_time before renewing.

**Comment by normanr** [comment]
Was there a plan to update `netinit_net_bringup` to use `dhcpc_request_async`? and/or add some built-in commands like `ntpc` has to start/stop a dhcpc daemon?

**Comment by xiaoxiang781216** [comment]
Nice idea, @coffee809721232 please take a look.

File: netutils/dhcpc/dhcpc.c (hunk 3 above)
Comment (xiaoxiang781216) [reply netutils/dhcpc/dhcpc.c:481]: @coffee809721232

Comment (coffee809721232) [reply netutils/dhcpc/dhcpc.c:481]: > Note: This should only sleep half of the lease_time before renewing.

Yes, it should be half of the time to re-request.
This bug will be fixed later.
thanks. @normanr @xiaoxiang781216

**Comment by coffee809721232** [comment]
> Was there a plan to update `netinit_net_bringup` to use `dhcpc_request_async`? and/or add some built-in commands like `ntpc` has to start/stop a dhcpc daemon?
Plan to add commands to start and stop the dhcpc daemon

File: netutils/dhcpc/dhcpc.c (hunk 3 above)
Comment (normanr) [reply netutils/dhcpc/dhcpc.c:481]: fixed in: https://github.com/apache/incubator-nuttx-apps/pull/958

**Comment by normanr** [comment]
fyi: I tried to use this code (adding a `dhcpc` daemon based on the `ntpc` daemon), but it has a couple of nasty bugs:
- it doesn't renew dhcp leases, but rather does full re-discovery and request (as if it had never seen the network before),
- you can't start `dhcpc_request_async` from `netinit_net_bringup`, and later call `dhcpc_close` from nsh:
  - the socket isn't closed properly, not sure why (it works fine when calling `dhcpc_request_async` from nsh),
- the interface name stored in `dhcpc_state_s` isn't duplicated, so the name is lost when the caller frees the underlying string,
- sometimes it sees a double-offer, it it will decline it even if it's just a duplicate of the offer that it's just received.

**Comment by xiaoxiang781216** [comment]
> fyi: I tried to use this code (adding a `dhcpc` daemon based on the `ntpc` daemon), but it has a couple of nasty bugs:
> 
> * it doesn't renew dhcp leases, but rather does full re-discovery and request (as if it had never seen the network before),

yes, we can optimize in this case.

> * you can't start `dhcpc_request_async` from `netinit_net_bringup`, and later call `dhcpc_close` from nsh:
>   
>   * the socket isn't closed properly, not sure why (it works fine when calling `dhcpc_request_async` from nsh),

does you call dhcpc_request_async and dhcpc_close from the different task?

> * the interface name stored in `dhcpc_state_s` isn't duplicated, so the name is lost when the caller frees the underlying string,

The design want to save one string copy, and is easy to use copy instead.

> * sometimes it sees a double-offer, it it will decline it even if it's just a duplicate of the offer that it's just received.

**Comment by normanr** [comment]
> > * you can't start `dhcpc_request_async` from `netinit_net_bringup`, and later call `dhcpc_close` from nsh:
> >   * the socket isn't closed properly, not sure why (it works fine when calling `dhcpc_request_async` from nsh),
> does you call dhcpc_request_async and dhcpc_close from the different task?

yes I figured that was why it breaks. I assume that all nsh builtins end up running in the same task, so that probably explains why it works. With more work it should be possible to refactor the code so that it's more like the ntpc daemon (with proper state synchronization) so that it can close the socket from the task that opened it.

Unlike the ntpc daemon, the dhcpc daemon should probably be able support multiple interfaces simultaneously. So it might just be easier to write it as a regular application (where the interface name is given on the command line) that can be safely killed with a signal. That would probably be a simpler way to solve the issue with closing the socket too.

> > * the interface name stored in `dhcpc_state_s` isn't duplicated, so the name is lost when the caller frees the underlying string,
> The design want to save one string copy, and is easy to use copy instead.

yea, it was easy to fix by adding `dup()` in `dhcpc_open` (and an extra `free()` before existing `free()` calls).

**Comment by normanr** [comment]
> * sometimes it sees a double-offer, it it will decline it even if it's just a duplicate of the offer that it's just received.

I noticed that this is _not_ how [RFC 2131](https://datatracker.ietf.org/doc/html/rfc2131) intends the client to behave. An `OFFER` is only supposed to be `DECLINED` "If the client detects that the address is already in use (e.g., through the use of ARP)" (and only at the `ACK` stage of processing. If multiple offers are received then the client indicates which server's offer it is accepting through the `server identifier` field.

I also discovered that if any other udp packets arrive at the client while it's sleeping (e.g. from _other_ clients on the same network doing dhcp requests), then they will consume IO buffers, and never give them up, because `dhcpc_request` doesn't expect responses to be queued at the socket. This ends up starving all other network connections on the system. Changing the sleep into a recv & discard loop fixes this.

**Pull request merged**